edx_sga/sga.py

```diff
@@ -0,0 +1,530 @@
+    def getter(inst):
+        value = meth(inst)
+        inst.__dict__[meth.__name__] = value
+        return value
+    return property(getter)
+
+
+class StaffGradedAssignmentXBlock(XBlock):
+    """
+    This block defines a Staff Graded Assignment.  Students are shown a rubric
+    and invited to upload a file which is then graded by staff.
+    """
+    has_score = True
+    icon_class = 'problem'
+
+    display_name = String(
+        default='Staff Graded Assignment', scope=Scope.settings,
+        help="This name appears in the horizontal navigation at the top of "
+             "the page."
+    )
+
+    weight = Float(
+        display_name="Problem Weight",
+        help=("Defines the number of points each problem is worth. "
+              "If the value is not set, the problem is worth the sum of the "
+              "option point values."),
+        values={"min": 0, "step": .1},
+        scope=Scope.settings
+    )
+
+    points = Integer(
+        display_name="Maximum score",
+        help=("Maximum grade score given to assignment by staff."),
+        default=100,
+        scope=Scope.settings
+    )
+
+    staff_score = Integer(
+        display_name="Score assigned by non-instructor staff",
+        help=("Score will need to be approved by instructor before being "
+              "published."),
+        default=None,
+        scope=Scope.settings
+    )
+
+    comment = String(
+        display_name="Instructor comment",
+        default='',
+        scope=Scope.user_state,
+        help="Feedback given to student by instructor."
+    )
+
+    annotated_sha1 = String(
+        display_name="Annotated SHA1",
+        scope=Scope.user_state,
+        default=None,
+        help=("sha1 of the annotated file uploaded by the instructor for "
+              "this assignment.")
+    )
+
+    annotated_filename = String(
+        display_name="Annotated file name",
+        scope=Scope.user_state,
+        default=None,
+        help="The name of the annotated file uploaded for this assignment."
+    )
+
+    annotated_mimetype = String(
+        display_name="Mime type of annotated file",
+        scope=Scope.user_state,
+        default=None,
+        help="The mimetype of the annotated file uploaded for this assignment."
+    )
+
+    annotated_timestamp = DateTime(
+        display_name="Timestamp",
+        scope=Scope.user_state,
+        default=None,
+        help="When the annotated file was uploaded")
+
+    def max_score(self):
+        return self.points
+
+    @reify
+    def block_id(self):
+        # cargo culted gibberish
+        return self.scope_ids.usage_id
+
+    def student_submission_id(self, id=None):
+        """
+        Returns dict required by the submissions app for creating and
+        retrieving submissions for a particular student.
+        """
+        if id is None:
+            id = self.xmodule_runtime.anonymous_student_id
+            assert id != 'MOCK', "Forgot to call 'personalize' in test."
```

Comment:
Seems a little odd to have test code bleed into this function. Maybe it's worth validating the id is properly set in the test's setUp function? Not a big deal, either way.